New tutorial for ROS 2 actions (#350)
ros2/ros2_documentation: source/Tutorials/UnderstandingROS2Actions.rst

When the server-side preempts an action, it “aborts” the goal.

Try hitting the ``D`` key, then the ``G`` key before the first rotation can complete.
In the terminal where the ``/turtlesim`` node is running, you will see the message:

.. code-block:: bash

  [WARN] [turtlesim]: Rotation goal received before a previous goal finished. Aborting previous goal

The server-side aborted the first goal because it was interrupted.

3 ros2 node info
^^^^^^^^^^^^^^^^

To see the ``/turtlesim`` node’s actions, open a new terminal and run the command:

.. code-block:: bash

    ros2 node info /turtlesim

Which will return a list of ``/turtlesim``’s subscribers, publishers, services, action servers and action clients:

.. code-block:: bash

  /turtlesim
    Subscribers:
      /parameter_events: rcl_interfaces/msg/ParameterEvent
      /turtle1/cmd_vel: geometry_msgs/msg/Twist
    Publishers:
      /parameter_events: rcl_interfaces/msg/ParameterEvent
      /rosout: rcl_interfaces/msg/Log
      /turtle1/color_sensor: turtlesim/msg/Color
      /turtle1/pose: turtlesim/msg/Pose

>       /turtle1/rotate_absolute/_action/feedback: turtlesim/action/RotateAbsolute_FeedbackMessage

Reviewer: Having these hidden names here reminded me to follow up and remove them from `ros2 node info` output by default. See https://github.com/ros2/ros2cli/pull/373

We can update the tutorial after this PR (and when the referenced PR lands) if you prefer.

Author: I think it landed already, right? I pulled the changes and now they're removed from `node info` output for me

Enter the following command with the action type in your terminal:

.. code-block:: bash

    ros2 interface show turtlesim/action/RotateAbsolute.action

Which will return:

.. code-block:: bash

  # The desired heading in radians
  float32 theta
  ---
  # The angular displacement in radians to the starting position
  float32 delta
  ---
  # The remaining rotation in radians
  float32 remaining

The first section of this message, above the ``---``, is the structure (data type and name) of the goal request.
The next section is the structure of the result.
The last section is the structure of the feedback.

7 ros2 action send_goal
^^^^^^^^^^^^^^^^^^^^^^^

Now let’s send an action goal from the command line with the following syntax:

.. code-block:: bash

    ros2 action send_goal <action_name> <action_type> <values>

``<values>`` need to be in YAML format.

Keep an eye on the turtlesim window, and enter the following command into your terminal:

.. code-block:: bash

    ros2 action send_goal /turtle1/rotate_absolute turtlesim/action/RotateAbsolute {'theta: 1.57'}

You should see the turtle rotating, as well as the following message in your terminal:

> .. code-block:: bash

Reviewer: Small nit: might be better to omit the `bash` suggestion for the code-block. Otherwise, it seems to highlight some arbitrary parts of the output being rendered (e.g. `for` in `Waiting for ...` is interpreted as a keyword)

.. _ROS2Actions:

Understanding ROS 2 actions
===========================

**Goal:** Introspect actions in ROS 2.

**Tutorial level:** Beginner

**Time:** 15 minutes

.. contents:: Contents
   :depth: 2
   :local:

Background
----------

Actions are one of the communication types in ROS 2 intended for long running tasks.
They consist of three parts: a goal, a result, and feedback.

Actions are built on topics and services.
Their functionality is similar to services, except actions are preemptable (you can cancel them while executing).

> They also provide steady feedback, as opposed to services which return a single response right away.

Reviewer: Services don't necessarily have to return right away.
```suggestion
They also provide steady feedback, as opposed to services which return a single response.
```

As always, don’t forget to source ROS 2 in **every new terminal you open**.

Tasks
-----

1 Setup
^^^^^^^

Start up the two turtlesim nodes, ``/turtlesim`` and ``/teleop_turtle``.

Open a new terminal and run:

.. code-block:: bash

    ros2 run turtlesim turtlesim_node

Open another terminal and run:

.. code-block:: bash

    ros2 run turtlesim turtle_teleop_key

2 Use actions
^^^^^^^^^^^^^

When you launch the ``/teleop_turtle`` node, you will see the following message in your terminal:

.. code-block:: bash

    Use arrow keys to move the turtle.
    Use G|B|V|C|D|E|R|T keys to rotate to absolute orientations. 'F' to cancel a rotation.

Let’s focus on the second line, which corresponds to an action.
(The first instruction corresponds to the “cmd_vel” topic, discussed previously in the topics tutorial.)

.. link topics tutorial

Notice that the letter keys ``G|B|V|C|D|E|R|T`` form a “box” around the ``F`` key on your keyboard.
Each key’s position around ``F`` corresponds to that orientation in turtlesim.
For example, the ``E`` will rotate the turtle’s orientation to the upper left corner.

Pay attention to the terminal where the ``/turtlesim`` node is running.

> Each time you press one of these keys, you are giving the action a goal to rotate.

Author: ```
...giving the action a goal...
```
Doesn't sound right?

A message relaying the result of the goal should display once the turtle completes its rotation:

.. code-block:: bash

    [INFO] [turtlesim]: Rotation goal completed successfully

The ``F`` key will cancel a goal mid-execution, demonstrating the preemptable feature of actions.

Try pressing the ``C`` key, and then pressing the ``F`` key before the turtle can complete its rotation.
In the terminal where the ``/turtlesim`` node is running, you will see the message:

.. code-block:: bash

  [INFO] [turtlesim]: Rotation goal canceled

Not only can the client-side (your input in the teleop) cancel goal, but the server-side (the ``/turtlesim`` node) can as well.

> When the server-side cancels an action, it “aborts” the goal.

Author: I think this comment is correct, but not sure if the language is accurate

        /kill: turtlesim/srv/Kill
        /reset: std_srvs/srv/Empty
        /spawn: turtlesim/srv/Spawn
        /turtle1/rotate_absolute/_action/cancel_goal: action_msgs/srv/CancelGoal
        /turtle1/rotate_absolute/_action/get_result: turtlesim/action/RotateAbsolute_GetResult
        /turtle1/rotate_absolute/_action/send_goal: turtlesim/action/RotateAbsolute_SendGoal
        /turtle1/set_pen: turtlesim/srv/SetPen
        /turtle1/teleport_absolute: turtlesim/srv/TeleportAbsolute
        /turtle1/teleport_relative: turtlesim/srv/TeleportRelative
        /turtlesim/describe_parameters: rcl_interfaces/srv/DescribeParameters
        /turtlesim/get_parameter_types: rcl_interfaces/srv/GetParameterTypes
        /turtlesim/get_parameters: rcl_interfaces/srv/GetParameters
        /turtlesim/list_parameters: rcl_interfaces/srv/ListParameters
        /turtlesim/set_parameters: rcl_interfaces/srv/SetParameters
        /turtlesim/set_parameters_atomically: rcl_interfaces/srv/SetParametersAtomically
    Actions Servers:
        /turtle1/rotate_absolute: turtlesim/action/RotateAbsolute
    Actions Clients:

.. todo: change this once hidden things are removed from the list

> Notice that the ``/turtle1/rotate_absolute`` action for ``/turtlesim`` is under ``Actions Servers``.

Reviewer: I think we should also update the tool so it reads "Action Servers" ("Action" is singular). Same for action clients.
I can open a PR with this change.

Reviewer: https://github.com/ros2/ros2cli/pull/357

This means ``/turtlesim`` responds to and provides feedback for the ``/turtle1/rotate_absolute`` action.

> The ``/teleop_turtle`` node has the ``/turtle1/rotate_absolute`` under ``Actions Clients`` because it executes the goal:

Author: Needs review

> The ``/teleop_turtle`` node has the ``/turtle1/rotate_absolute`` under ``Actions Clients`` because it executes the goal:

Reviewer: The client isn't *executing* goals, but rather *sending* them.
```suggestion
The ``/teleop_turtle`` node has the name ``/turtle1/rotate_absolute`` under ``Action Clients`` meaning that it sends goals for that action name.
```

.. code-block:: bash

    ros2 node info /teleop_turtle

Which will return:

.. code-block:: bash

  /teleop_turtle
    Subscribers:
        /parameter_events: rcl_interfaces/msg/ParameterEvent
        /turtle1/rotate_absolute/_action/feedback: turtlesim/action/RotateAbsolute_FeedbackMessage
        /turtle1/rotate_absolute/_action/status: action_msgs/msg/GoalStatusArray
    Publishers:
        /parameter_events: rcl_interfaces/msg/ParameterEvent
        /rosout: rcl_interfaces/msg/Log
        /turtle1/cmd_vel: geometry_msgs/msg/Twist
    Services:
        /teleop_turtle/describe_parameters: rcl_interfaces/srv/DescribeParameters
        /teleop_turtle/get_parameter_types: rcl_interfaces/srv/GetParameterTypes
        /teleop_turtle/get_parameters: rcl_interfaces/srv/GetParameters
        /teleop_turtle/list_parameters: rcl_interfaces/srv/ListParameters
        /teleop_turtle/set_parameters: rcl_interfaces/srv/SetParameters
        /teleop_turtle/set_parameters_atomically: rcl_interfaces/srv/SetParametersAtomically
    Actions Servers:

    Actions Clients:
        /turtle1/rotate_absolute: turtlesim/action/RotateAbsolute

4 ros2 action list
^^^^^^^^^^^^^^^^^^

> To identify all the actions in your current system, run the command:

Reviewer: I'm not sure if the concept of a "ROS graph" has already been introduced in a previous tutorial, but I think it is more concrete use that term:
```suggestion
To identify all the actions in the ROS graph, run the command:
```

.. code-block:: bash

    ros2 action list

Which will return:

.. code-block:: bash

    /turtle1/rotate_absolute

> This is the only action in the system right now.

Reviewer: ```suggestion
This is the only action in the ROS graph right now.
```

It controls the turtle’s rotation, as you saw earlier.

> You also already know that this action is an action client for ``/teleop_turtle`` and an action server for ``/turtlesim`` from using the ``ros2 node info <node_name>`` command.

Reviewer: I wouldn't say that "an action is a client or server". It's more like action clients and servers are associated with a name that identifies the channel in for communication. Analogous to topics for pub/subs.
```suggestion
You also already know that there is one action client (part of ``/teleop_turtle``) and one action server (part of ``/turtlesim``) for this action from using the ``ros2 node info <node_name>`` command.
```

4.1 ros2 action list -t
~~~~~~~~~~~~~~~~~~~~~~~

Actions have types, similar to topics and services.
To find ``/turtle1/rotate_absolute``'s type, run the command:

.. code-block:: bash

    ros2 action list -t

Which will return:

.. code-block:: bash

    /turtle1/rotate_absolute [turtlesim/action/RotateAbsolute]

In brackets to the left of each action name (in this case only ``/turtle1/rotate_absolute``) is the action type, ``turtlesim/action/RotateAbsolute``.

> You will need this when you want to execute an action from the command line.

Reviewer: ```suggestion
You will need this when you want to execute an action from the command line or from code.
```

5 ros2 action info
^^^^^^^^^^^^^^^^^^

You can further introspect the ``/turtle1/rotate_absolute`` action with the command:

.. code-block:: bash

    ros2 action info /turtle1/rotate_absolute

Which will return

.. code-block:: bash

  Action: /turtle1/rotate_absolute
  Action clients: 1
      /teleop_turtle
  Action servers: 1
      /turtlesim

This tells us what we learned earlier from running ``ros2 node info`` on each node:

> ``/turtle1/rotate_absolute`` is an action client for the ``/teleop_turtle`` node, and an action server for the ``/turtlesim`` node.

Reviewer: Similar to my comment above. I think of the relationship between nodes, action clients/servers, and actions a little differently.
```suggestion
the ``/teleop_turtle`` node has an action client and the ``/turtlesim`` node has an action server for the ``/turtle1/rotate_absolute`` action.
```

6 ros2 action show
^^^^^^^^^^^^^^^^^^

> One more piece of information you will need before executing an action goal yourself is the structure of the action type.

Reviewer: ```suggestion
One more piece of information you will need before sending or executing an action goal yourself is the structure of the action type.
```

Recall that you identified ``/turtle1/rotate_absolute``’s type when running the command ``ros2 action list -t ``.
Enter the following command with the action type in your terminal:

.. code-block:: bash

    ros2 action show turtlesim/action/RotateAbsolute

Which will return:

.. code-block:: bash

  # The desired heading in radians
  float32 theta
  ---
  # The angular displacement in radians to the starting position
  float32 delta
  ---
  # The remaining rotation in radians
  float32 remaining

> The first section of this message, above the ``---``, is the structure (data type and name) of the goal input.

Author: ```
...goal input.
```
Doesn't sound right

Reviewer: In code, we refer to it as the "goal request" or simply "goal" depending on the context.

The next section is the structure of the result.
The last section is the structure of the feedback.

7 ros2 action send_goal
^^^^^^^^^^^^^^^^^^^^^^^

> Now let’s execute an action goal from the command line with the following syntax:

Reviewer: Again, I'd avoid the word *execute* in this context, since to me it implies actually carrying out the goal versus making a request.
```suggestion
Now let’s send an action goal from the command line with the following syntax:
```

You can also see the result, a field with the name ``delta``, which is the displacement to the starting position.

To see the feedback of this goal, add ``--feedback`` to the last command you ran.
First, make sure you change the value of ``theta``.
After running the previous command, the turtle will already be at the orientation of ``1.57`` radians, so it won’t move unless you pass a new ``theta``.

.. code-block:: bash

    ros2 action send_goal /turtle1/rotate_absolute turtlesim/action/RotateAbsolute {'theta: -1.57'} --feedback

Your terminal will return the message:

.. code-block:: bash

  Sending goal:
     theta: -1.57

  Goal accepted with ID: e6092c831f994afda92f0086f220da27

  Feedback:
    remaining: -3.1268222332000732

  Feedback:
    remaining: -3.1108222007751465

  …

  Result:
    delta: 3.1200008392333984

  Goal finished with status: SUCCEEDED

You will continue to receive feedback, the remaining radians, until the goal is complete.

Summary
-------

Actions are like services that allow you to execute long running tasks, provide regular feedback, and are cancelable.

A robot system would likely use actions for navigation.

> An action could tell a robot to travel a position.

Reviewer: ```suggestion
An action goal could tell a robot to travel to a position.
```

> The goal would navigate and send updates along the way, and then a final result message once it's reached its destination.

Reviewer: ```suggestion
While the robot navigates to the position it can send updates along the way (i.e. feedback), and then a final result message once it's reached its destination.
```

> Turtlesim has one action that acts as a client for one node and server for the other.

Reviewer: ```suggestion
Turtlesim has an action server that action clients can send goals to for rotating turtles.
```
